libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/ActivityChecker.cs

```diff
@@ -0,0 +1,372 @@
+                    var value = item.Value.ToString().ToLower();
+
+                    switch (property.ToLower())
+                    {
+                        case "$type":
+                        case "title":
+                        case "value":
+                        case "displaytext":
+                        case "text":
+                        case "image":
+                            break;
+
+                        case "type":
+                            if (value != ActionTypes.ImBack.ToLower()
+                                && value != ActionTypes.Call.ToLower()
+                                && value != ActionTypes.DownloadFile.ToLower()
+                                && value != ActionTypes.MessageBack.ToLower()
+                                && value != ActionTypes.openApp.ToLower()
+                                && value != ActionTypes.OpenUrl.ToLower()
+                                && value != ActionTypes.Payment.ToLower()
+                                && value != ActionTypes.PlayAudio.ToLower()
+                                && value != ActionTypes.PlayVideo.ToLower()
+                                && value != ActionTypes.PostBack.ToLower()
+                                && value != ActionTypes.ShowImage.ToLower()
+                                && value != ActionTypes.Signin.ToLower())
+                            {
+                                result.Add(BuildDiagnostic($"'{value}' is not a valid action type"));
+                            }
+
+                            break;
+                        default:
+                            result.Add(BuildDiagnostic($"'{property}' is not support for card action.", false));
+                            break;
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private static string GetStructureType(JObject jObj)
+        {
+            if (jObj == null)
+            {
+                return string.Empty;
+            }
+
+            var type = jObj["$type"]?.ToString()?.Trim();
+            if (string.IsNullOrEmpty(type))
+            {
+                // Adaptive card type
+                type = jObj["type"]?.ToString()?.Trim();
+            }
+
+            return type?.ToLower() ?? string.Empty;
+        }
+
+        private static List<Diagnostic> CheckAttachments(JToken value)
+        {
+            var result = new List<Diagnostic>();
+
+            var attachmentsJsonList = NormalizedToList(value);
+
+            foreach (var attachmentsJson in attachmentsJsonList)
+            {
+                if (attachmentsJson is JObject attachmentsJsonJObj)
+                {
+                    result.AddRange(CheckAttachment(attachmentsJsonJObj));
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckAttachment(JObject lgJObj)
+        {
+            var result = new List<Diagnostic>();
+
```

Comment:
All of this formatting needs to adhere with our rules.

Reply:
May I ask which rule this code does not adhere? thanks